timemachine/cpp/src/segmented_logsumexp.hpp

```diff
@@ -0,0 +1,34 @@
+#pragma once
+
+#include "device_buffer.hpp"
+
+namespace timemachine {
+
+template <typename RealType> class SegmentedLogSumExp {
+
+private:
+    const int max_vals_per_segment_;
+    const int num_segments_;
+    DeviceBuffer<RealType> d_temp_buffer_;
+    // Buffers for finding the max/summed value
+    std::size_t temp_storage_bytes_;
+    DeviceBuffer<char> d_temp_storage_buffer_;
+
+public:
+    SegmentedLogSumExp(const int max_vals_per_segment, const int num_segments);
+
+    ~SegmentedLogSumExp();
+
+    void sum_device(
```

Comment:
The reason for this is that to combine them would require running an additional kernel that is a trivial operation better pushed into other kernels. Will add back in a comment, was in the previous LogSumExp but removed here.

Reply:
Attempted to clarify this with comments in https://github.com/proteneer/timemachine/pull/1239/commits/0743f58dfb303e8ed4b3929398594b52a7d613c8

Reply:
>  trivial operation better pushed into other kernels

Some considerations:

0) this is at the cost of having double buffers for `(d_lse_max_before_, d_lse_exp_sum_before_), (d_lse_max_after_, d_lse_exp_sum_after_)`, you may still end-up paying for it in terms of additional memsets etc. required. 
1) unless other downstream calls (outside of this class) depend on the individual components (max, exp sum) in a way that's outside of the logsumexp and you've benchmarked it, i'd strongly prefer to encapsulate the lse functionality inside the class as much as possible. 

wdyt?

Reply:
> this is at the cost of having double buffers

Yes, you do pay 2x the memory usage, there are no memsets around these buffers so there is no additional computational cost.

> i'd strongly prefer to encapsulate the lse functionality inside the class as much as possible.

Think that is reasonable, though now we are going from batching the old implementation (which did this same thing, just done in one array since easy to offset without the segments, already a deboggling comment to figure out how to do that in CUB - https://github.com/proteneer/timemachine/blob/0743f58dfb303e8ed4b3929398594b52a7d613c8/timemachine/cpp/src/segmented_logsumexp.cu#L53) to modifying the implementation. Not clear of what utility this adds for our immediate goals and would rather leave that until we have sorted out batching the exchange moves more broadly.

Reply:
Thinking about this a bit more, haven't actually benchmarked in a while (too early on the exchange move stuff to be concrete), think this would add quite a bit of overhead.

If you look at the current implementation (From PR: https://github.com/proteneer/timemachine/pull/1225) which has `k_exp_sub_max` which takes 1.7k ns per kernel call which is run twice for every TIBD proposal. Despite effectively the lowest run time possible (k_arange is 1.7k ns as well), because it is run thousands of time the kernel takes up 1.8% of the entire run time.

```
Time (%)  Total Time (ns)  Instances   Avg (ns)     Med (ns)    Min (ns)   Max (ns)   StdDev (ns)                                                  Name                                                
 --------  ---------------  ---------  -----------  -----------  ---------  ---------  -----------  ----------------------------------------------------------------------------------------------------
1.8       44,713,131     25,001      1,788.5      1,792.0      1,727      2,208         50.6  void timemachine::k_exp_sub_max<float>(int, const T1 *, const T1 *, T1 *)
```

Even assuming that the new kernel to combine them takes half that long, would be almost a percent of the entire run time.

Reply:
Discussed offline, this was what was done previously. Need to follow up with

* Rename `SegmentedLogSumExp` -> `SegmentedSumExp` and document how to compute `logsumexp`

Reply:
Renamed SegmentedLogSumExp to SegmentedSumExp to clarify that the device code does not compute LogSumExp directly, includes comment pointing to code to compute the final logsumexp value. https://github.com/proteneer/timemachine/pull/1239/commits/f512f73e4d72ca61fc4f1e433c3af0cf291345a3